package_scanner.nim

```diff
@@ -1,194 +1,218 @@
+  var headers: HttpHeaders = nil
+  if url.startsWith("https://github.com"):
+    if existsEnv("GITHUB_TOKEN"):
+      headers = newHttpHeaders({"Authorization": "Bearer " & getEnv("GITHUB_TOKEN")})
+
+  try:
+    let resp = client.request(url, headers=headers)
+    discard resp.bodyStream.readAll()
+    if not resp.code.is2xx:
+      result = "Server returned status " & $resp.code
+  except TimeoutError:
+    result = "Timeout after " & $client.timeout & "ms"
+    client.close()
+  except HttpRequestError:
+    result = "HTTP error: " & getCurrentExceptionMsg()
+    client.close()
+  except AssertionDefect:
+    result = "httpclient error: " & getCurrentExceptionMsg()
+    client.close()
+  except CatchableError as e:
+    result = "Unexpected exception " & $e.name & ": " & e.msg
+    client.close()
+
+template logPackageError(errorMsg: string) =
+  echo "E: ", errorMsg
+  success = false
+
+template checkUrl(urlType: string, url: string) =
+  if url == "":
+    logPackageError(displayName & " has an empty " & urlType & " URL")
+  elif not url.startsWith("https://"):
+    logPackageError(displayName & " has a non-https " & urlType & " URL: " & url)
+  elif checkUrls:
+    let urlError = client.checkUrlReachable(url)
+    if urlError != "":
+      logPackageError(displayName & " has an unreachable " & urlType & " URL: " & url)
+      logPackageError(urlError)
+
+proc getStrIfExists(n: JsonNode, name: string, default: string = ""): string =
+  result = default
+  if n.hasKey(name) and n[name].kind == JString:
+    result = n[name].str
+
+proc getElemsIfExists(n: JsonNode, name: string, default: seq[JsonNode] = @[]): seq[JsonNode] =
+  result = default
+  if n.hasKey(name) and n[name].kind == JArray:
+    result = n[name].elems
+
+proc checkPackages(newPackagesPath: string, oldPackagesPath: string, checkUrls: bool = false): int =
+  var oldPackagesTable = initTable[string, JsonNode]()
+  if oldPackagesPath != "":
+    let oldPackagesJson = parseJson(readFile(oldPackagesPath))
+    for oldPkg in oldPackagesJson:
+      let oldNameLower = oldPkg.getStrIfExists("name").toLowerAscii()
+      if oldNameLower != "":
+        oldPackagesTable[oldNameLower] = oldPkg
+
+  let newPackagesJson = parseJson(readFile(newPackagesPath))
+  # Do a first pass through the list to count duplicate names
+  var packageNameCounter = initCountTable[string]()
+  for pkg in newPackagesJson:
+    let pkgNameLower = pkg.getStrIfExists("name").toLowerAscii()
+    if pkgNameLower != "":
+      packageNameCounter.inc(pkgNameLower)
+
+  var client: HttpClient = nil
+  if checkUrls:
+    client = newHttpClient(timeout=3000)
+    client.headers = newHttpHeaders({"User-Agent": "Nim packge_scanner/2.0"})
+
+  var modifiedPackages = 0
+  var failedPackages = 0
+  for pkg in newPackagesJson:
+    var success = true  # Set to false by logPackageError
+    let pkgName = pkg.getStrIfExists("name")
+    let pkgNameLower = pkgName.toLowerAscii()
+    var displayName = pkgName
```

Comment:
I did this to be able to report all errors in one pass. This means packages do not have to be resubmitted with a valid name to get the remaining errors. It is a small benefit but since I wanted to consistently remove all early fails from the logic it only made sense to do it this way.

Reply:
Yeah, it shouldn't block this PR. Feel free to change the logic later.